rest/handler.go

```diff
@@ -290,14 +288,18 @@ func (h *handler) validateAndWriteHeaders(method handlerMethod, accessPermission
 		}
 	}()
 
+	h.requestBody = NewReaderCounter(h.rq.Body)
+
 	switch h.rq.Header.Get("Content-Encoding") {
 	case "":
-		h.requestBody = h.rq.Body
+		h.requestBody = NewReaderCounter(h.rq.Body)
 	case "gzip":
+		var gzipRequestReader *gzip.Reader
 		var err error
-		if h.requestBody, err = gzip.NewReader(h.rq.Body); err != nil {
+		if gzipRequestReader, err = gzip.NewReader(h.rq.Body); err != nil {
```

Comment:
Nit: Could avoid the `var` declarations if we drop the error handling out of the assignments.

```suggestion
		gzipRequestReader, err := gzip.NewReader(h.rq.Body)
		if err != nil {
```